lighthouse-core/audits/valid-source-maps.js

```diff
@@ -0,0 +1,162 @@
+  title: 'Page has valid source maps',
+  /** Title of a Lighthouse audit that provides detail on HTTP to HTTPS redirects. This descriptive title is shown to users when HTTP traffic is not redirected to HTTPS. */
+  failureTitle: 'Missing source maps for large first-party JavaScript',
+  /** Description of a Lighthouse audit that tells the user that their JavaScript source maps are invalid or missing. This is displayed after a user expands the section to see more. No character length limits. 'Learn More' becomes link text to additional documentation. */
+  description: 'Source maps translate your minified code to the original source code which ' +
+    'helps with debugging, and tracking down bugs in production. Consider enabling or updating ' +
+    'source maps during JavaScript bundling to take advantage of this feature. ' +
+    '[Learn more](https://developers.google.com/web/tools/chrome-devtools/javascript/source-maps).',
+};
+
+const str_ = i18n.createMessageInstanceIdFn(__filename, UIStrings);
+
+const LARGE_JS_BYTE_THRESHOLD = 500 * 1024;
+
+/**
+ * Returns true if the script URL is either:
+ * (1) a known third-party script or
+ * (2) the same as the (redirected) requested Lighthouse URL
+ * @param {string} url
+ * @param {string} finalUrl
+ */
+function isFirstParty(url, finalUrl) {
+  try {
+    const entity = thirdPartyWeb.getEntity(url);
+    if (!entity) return true;
+    return entity === thirdPartyWeb.getEntity(finalUrl);
+  } catch (_) {
+    return false;
+  }
+}
+
+class ValidSourceMaps extends Audit {
+  /**
+   * @return {LH.Audit.Meta}
+   */
+  static get meta() {
+    return {
+      id: 'valid-source-maps',
+      title: str_(UIStrings.title),
+      failureTitle: str_(UIStrings.failureTitle),
+      description: str_(UIStrings.description),
+      requiredArtifacts: ['ScriptElements', 'SourceMaps', 'URL'],
+    };
+  }
+
+  /**
+   * Returns true if the size of the script exceeds a static threshold
+   * @param {LH.Artifacts.ScriptElement} scriptElement
+   * @param {string} finalURL
+   * @return {boolean}
+   */
+  static isLargeFirstPartyJS(scriptElement, finalURL) {
+    if (scriptElement.content === null) return false;
+
+    const isLargeJS = scriptElement.content.length >= LARGE_JS_BYTE_THRESHOLD;
+    const isFirstPartyJS = scriptElement.src ? isFirstParty(scriptElement.src, finalURL) : false;
+
+    return isLargeJS && isFirstPartyJS;
+  }
+
+  /**
+   * @param {LH.Artifacts} artifacts
+   * @param {LH.Audit.Context} _
+   */
+  static async audit(artifacts, _) {
+    const {SourceMaps} = artifacts;
+
+    /** @type {Set<string>} */
+    const isMissingMapForLargeFirstPartyScriptUrl = new Set();
+
+    let missingMapsForLargeFirstPartyFile = false;
+    const results = [];
+    for (const ScriptElement of artifacts.ScriptElements) {
+      if (!ScriptElement.src) continue; // TODO: inline scripts, how do they work?
```

Comment:
This was a note I took because I need to go through all uses of our source maps and see what happens if the script is inline with a source map url comment. I think many usages assume a network record exists, and I'm not sure what happens when that network record is the HTML document. I don't think inline bundled code is common ... so the work involved may just be to ensure it doesn't blow up Lighthouse. if supporting it would be too much work we won't bother.

I need to file an issue, then we can link to it here.

Reply:
> This was a note I took because I need to go through all uses of our source maps and see what happens if the script is inline with a source map url comment.

What does it mean for the script to be inline with a source map url comment?

> I don't think inline bundled code is common ... so the work involved may just be to ensure it doesn't blow up Lighthouse.

In this case, would an error be caught within the audit or break the entire runner?

Reply:
> What does it mean for the script to be inline with a source map url comment?

```
<script>
a;b;f;d;a;a;s;s;minified code.....

// sourceMapURL=file.js.map
</script>
```

> In this case, would an error be caught within the audit or break the entire runner?

I'm not sure, that's the investigation that this comment is alluding needs to be done. first of all, here L82 `const sourceMap =` wouldn't work because there is not a url to tie together a  source map and an inline script. I think the source map gatherer may not even process inline scripts?

in any case, it's not a high priority concern to support. I'd just want to see what happens w/ an inline script w/ a map, dbl check none of these new audits error.